tools/scripts/uconfig_vars_test.py

```diff
@@ -0,0 +1,217 @@
+# © 2021 and later: Unicode, Inc. and others.
+# License & terms of use: http://www.unicode.org/copyright.html
+
+"""Executes uconfig variations check.
+
+See
+http://site.icu-project.org/processes/release/tasks/healthy-code#TOC-Test-uconfig.h-variations
+for more information.
+"""
+
+import getopt
+import os
+import re
+import subprocess
+import sys
+
+excluded_unit_test_flags = ['UCONFIG_NO_CONVERSION', 'UCONFIG_NO_FILE_IO'];
+
+def ReadFile(filename):
+    """Reads a file and returns the content of the file
+
+    Args:
+      command: string with the filename.
+
+    Returns:
+      Content of file.
+    """
+
+    with open(filename, 'r') as file_handle:
+      return file_handle.read()
+
+def RunCmd(command):
+    """Execute the command, returns output and exit code, writes output to log
+
+    Args:
+      command: string with the command.
+
+    Returns: stdout and exit code of command execution.
+    """
+
+    command += ' >> uconfig_test.log 2>&1'
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE,
+                         stderr=subprocess.STDOUT, close_fds=True)
+    stdout, _ = p.communicate()
+    return stdout, p.returncode
+
+def SetUpICU():
+    """Configuration, installation of ICU4C."""
+
+    RunCmd('mkdir /tmp/icu_cnfg')
+    out, exit_code = RunCmd('./runConfigureICU Linux --prefix=/tmp/icu_cnfg')
+    if exit_code != 0:
+        print('ICU4C configuration failed!')
+        print(out)
+        sys.exit(-1)
+    _, exit_code = RunCmd('make -j2 install')
+    if exit_code != 0:
+        print('make install failed!')
+        sys.exit(-1)
+    # The header test is set up in a way that it uses uconfig.h in /tmp/icu_cnfg
+    # while the ICU4C unit tests uses common/unicode/uconfig.h. Set a link.
+    RunCmd('rm /tmp/icu_cnfg/include/unicode/uconfig.h')
+    os.symlink(os.path.abspath('common/unicode/uconfig.h'),
+               '/tmp/icu_cnfg/include/unicode/uconfig.h')
+    print('ICU setup completed.')
+
+def ExtractUConfigNoXXX(uconfig_file):
+    """Parses uconfig.h and returns a list of UCONFIG_NO_XXX labels.
+    Initializes test result structure.
+
+    Args:
+      uconfig_file: common/unicode/uconfig.h as string.
+
+    Returns:
+      List of all UCONFIG_NO_XXX flags found in uconfig.h, initialized test
+      result structure.
+    """
+
+    uconfig_no_flags_all = []
+    test_results = {}
+    uconfig_no_regex = r'UCONFIG_NO_[A-Z_]*'
+
+    for uconfig_no_flag in re.finditer(uconfig_no_regex, uconfig_file):
+        if uconfig_no_flag.group(0) not in uconfig_no_flags_all:
+            uconfig_no_flags_all.append(uconfig_no_flag.group(0))
+
+    test_results = {f: {'unit_test': False, 'hdr_test': False} for f in uconfig_no_flags_all}
+    test_results['all_flags'] = {'unit_test': False, 'hdr_test' : False}
+
+    return uconfig_no_flags_all, test_results
+
+def EnableUConfigNo(uconfig_file, uconfig_no_flag):
+    """Sets the uconfig_no_flag in common/unicode/uconfig.h file.
```

Comment:
It should not be necessary to rewrite uconfig.h. The config switches are all guarded by "if not defined, then define turnoff=0". So instead it would be simpler and cleaner (and require much less code) to just specify the relevant switch as a `-Dflag=1` option on the configure command line.

The symlink hack should also not be necessary.

Reply:
Underlying reason for doing it like this is runtime savings. Currently runConfigureICU and 'make install' are executed once only during setup.

Reply:
Hm. I didn't appreciate/remember that the hdrtst runs after `make install`.

Looking at the various Makefile.in files, each of the three libraries has an "install-headers" target. So for running hdrtst, it should suffice to configure, and then, for each of common|i18n|io you should be able to cd there and `make install-headers`. Then run the hdrtst. No need to actually build and install the whole libraries, which should also be much faster. And if that all is fast enough, you should be able to configure for each uconfig_no_something flag, without rewriting uconfig.h.

And yes, for running unit tests, you should not need to `make install`.

I don't remember if you already `make clean` between flags -- please make sure you do.

Reply:
The 'make install' is in the instructions (https://sites.google.com/site/icusite/processes/release/tasks/healthy-code?pli=1#TOC-Test-ICU4C-headers). I once tried to do without, had problems, and just continued as instructed.

Reply:
Changed, the UCONFIG_NO_XXX flag is no passed in via runConfigureICU (unit tests) or test/hdrtst/Makefile; test/hdrtst/Makefile.in adjusted accordingly.